classes/auth.php

```diff
@@ -962,6 +968,21 @@ public function get_email_from_attributes(array $attributes) {
         return false;
     }
 
+    /**
+     * Get lowercase username from attributes, force as lowercase because Moodle requires it.
+     *
+     * @param array $attributes A list of attributes.
+     *
+     * @return bool
+     */
+    private function get_username_from_attributes(array $attributes) {
+        if (!empty($this->config->field_map_username) && !empty($attributes[$this->config->field_map_username])) {
+            return strtolower($attributes[$this->config->field_map_username][0]);
```

Comment:
Could this ever be a string not an array in the SAML response?

Reply:
good thought - I traced through the code to see where it would usually set the username value in the update_user call and found:
https://github.com/catalyst/moodle-auth_saml2/blob/MOODLE_39_STABLE/classes/auth.php#L924

Which shows it's always treated as an array in the existing code, so we can leave this for now and come back to it later if needed.